### ISSUE TITLE
Heading variants with block and section space

### ISSUE THREAD

**Issue by lyubomir-popov** [canonical/vanilla-framework#4797]
The amount of space needed under a heading depends on context. Compare the following examples. The denser versions are the current default, which we can't change. Notice that this spacing looks fine when followed by a small paragraph. But they look inadequate when followed by a large block of text:

![image](https://github.com/canonical/vanilla-framework/assets/2741678/82819794-101a-4850-a561-78f4489e3b45)

When followed by large blocks of text or other visually "heavy" elements, block (1.5rem) or in this case section (4rem) spacing improves things significantly:

![image](https://github.com/canonical/vanilla-framework/assets/2741678/bad357ec-956c-41f2-bf16-5a173385b886)

But on the other hand, this can't be a default as it won't look good when followed by a single small paragraph:
![image](https://github.com/canonical/vanilla-framework/assets/2741678/a362601d-1375-4254-b241-fc0fb0c47252)

Currently, we can achieve this by wrapping headings  in `.p-section` or `.p-block`. @anthonydillon rightly questioned this approach, as it adds a lot of extra divs where a class would do. 

How about adding a class or modifier to achieve this, so it doesn't require extra divs?

cc @bartaz @anthonydillon

**Comment by bartaz** [comment]
I'm not sure what does it solve.

Currently we have a single purpose component `p-block` that is supposed to wrap around blocks of content that is a "sub-section" of a section. I'm quite surprised this is needed and used around single headings, as this is not how I understood this, but let's say we need to do this.

Currently to achieve this you do `<div class="p-block"><h2>Something</h2></div>`

As I understand the proposal is something like: `<h2 class="p-heading--2-block">Something</h2>`, or `<h2 class="is-block">` … I don't feel like adding another class names that do the same as p-block is any helpful. Yes, we save on a div, but now we have multiple components/classes that do the same thing: add "block" subsection spacing.

I believe underlying issue is not with the implementation, is it a component, or a variant, is just a detail. The point is how are devs supposed to know when to add the spacing around heading and when not to? It doesn't seem to be clear from Figma that doesn't use the same section/block components, it just seems to be arbitrarly positioned content, it's not easy to parse what components are there. And how are we going to control this "contextuality"? If someone changes number of paragraphs of text, are they now supposed to change spacing of heading above it? 

I find it hard to find any other design system that would be doing something similar (having headings with different spacing based on context). I'm not sure how to properly define it on Design System level to make it clear for everyone: designers and developers how are these supposed to work. Right now it seems that you have these rules and you apply them to your designs @lyubomir-popov but we need to "encode" them in a system that everyone could use consistently. 

Right now it feels like we are adding a lot of "contextual" arbitrary design exceptions here and there, and if we go this direction we are better off with Vanilla being like Tailwind rather than with predefined components. 

Long story short, I feel that if the proposed solution is adding more variants of headings it's a red flag that we have a deeper problem somewhere on higher level in the design system.

**Comment by lyubomir-popov** [comment]
> the point is how are devs supposed to know when to add the spacing around heading and when not to?

Other design systems use spacing presets, which they specify in designs. This way, the developer looks at the design and simply implements what has been requested there. We are not that far off with the new approach - we just have simpler presets, 24px and 64 px or block and section.

Maybe we could adopt this practice in mockups going forward - I don't mind adding blocks as in the  example here from IBM Carbon:
![image](https://github.com/canonical/vanilla-framework/assets/2741678/045e0c30-d6b7-4893-8a17-bfd2c04512ea)

We can go further and outline examples in the design system documentation, but any time the content change leads to adding/removing more than 20% content, a designer should review and flag if an intervention is needed.

> I find it hard to find any other design system that would be doing something similar (having headings with different spacing based on context).

Again, in the exmaple above, headings in IBM carbon have a minimum amount of default spacing, then at the designer's discretion space is specified in the mockup. This way, the designer is the one worrying about the context, not the developer.

---
In general, and I know this is a highly contentious notion, so I welcome input from others - the notion that a design should look great regardless of how much you alter the volume of content - I think this is not realistic.  In my experience, in all design disciplines including UX, UI and Brand - the volume of content is always a key consideration. This isn't a sign of a failing design system - this is simply aknowledging the importance of context. What do you think about that @juanruitina @cassiocassio @marcushas?

Here's an example: Imagine a list of elements - machines in maas, search results on a brochure sites, or even just files in a folder.
This would be a tipical decision process that UX/UI would go through  when deciding how to present the content:
- less than  5 - you can afford to display lots of detail for each item, for example in large cards
- between 5 and 50 - maybe you need pagination
- over 50 - you start thinking about tables, slicing and dicing mechanisms like search and filtering
- over 50,000 = you start thinking about dashboards with different overviews that summarise the information

We do very similar choices when designing a simple text page, but our "components" are not tables, paginations, search fields etc - they are simply blocks of white space and column arrangements.

**Comment by bartaz** [comment]
> Again, in the exmaple above, headings in IBM carbon have a minimum amount of default spacing, then at the designer's discretion space is specified in the mockup. This way, the designer is the one worrying about the context, not the developer.

The example you show is a component. In this case, in Modal it was decided that the spacing is like that. In another context/component it may be different.

The issue is, in our case, we don't have any "components" defined. We have headings that vary in spacing in arbitrary ways in same context: brochure sites.

So to follow the example of IBM Carbon we would need to define "content" components (strip with heading and text, strip with heading image text, etc). And on that level we can define what spacing is.

What we currently do is we treat every single page separately and design and build them from scratch without any reusable blocks outside of quite low level Vanilla styling.

Maybe the work around page templates that Juan and Sophie have been doing can help with that. If we have a predefined templates and "higher level" components for marketing pages, we could define these spacings on that level.

So far, I feel like the implementation we have in Vanilla is correct. We have a block component, that represents a sub section, and I think wrapping a heading with that if necessary is fine. I believe adding additional variants to heading components will just create more confusion.

**Comment by lyubomir-popov** [comment]
That's fine by me. Still, within a component, if you go from 5 words in a paragraph to 150, you'd need a variant of  that component with different spacing. Again, with or without components, context matters.

**Comment by bartaz** [comment]
> Still, within a component, if you go from 5 words in a paragraph to 150, you'd need a variant of that component with different spacing. Again, with or without components, context and volume matters.

I guess this could be addressed on template level, that for example we expect only 1 paragraph of text here, but the other place requires at least 3, or something like that. At least that's what I imagine a page template would do.

**Comment by lyubomir-popov** [comment]
That works for me. And as a general rule, any copy update that requires more than 20% change in content length needs to go through a design review.

**Comment by lyubomir-popov** [comment]
Maybe we can have a mini-sprint after we deliver c.com, to audit the new pages and break things down into components. At that time we'll have a large enough sample of variants to outline the numbmer of components needed and what variants each requires.

**Comment by marcushas** [comment]
@lyubomir-popov I think having a sprint to iron out the problems we are facing together sounds like a good plan

**Comment by bartaz** [comment]
We will have more flexible container component in place of section/block which should cover this cases #4809.

Still, designers will need to know when to use those around headings, and developers will need to be able to see it clearly in the designs (Figma).
